### ISSUE TITLE
Setting recipient email on checkout shipping error

### ISSUE THREAD

**Issue by james-allan** [Prospress/woocommerce-subscriptions-gifting#209]
While working on 3.0 compatibility I stumbled across this error: 

 ```
PHP Notice:  Undefined index: 2017_09_01_monthly_Insearruckin50@dayrep.com_0 in /wp-content/plugins/woocommerce-subscriptions/includes/class-wc-subscriptions-cart.php on line 1119
```

which occurred when I entered a recipient on the checkout and then tried to checkout. 

At first glance I imagine it's related to the fact that adding a recipient on checkout causes the recipient's email (in this case Insearruckin50@dayrep.com) to be appended to the recurring cart item key before subscriptions validates the shipping methods. This would cause a mismatch in the selected shipping methods (which are indexed by the recurring cart keys) and recurring cart item keys because the recurring cart keys were changed but the selected shipping methods weren't.

Another error logged at the same time is: 

```
PHP Notice:  Indirect modification of overloaded property WC_Checkout::$shipping_methods has no effect in /wp-content/plugins/woocommerce-subscriptions/includes/class-wc-subscriptions-cart.php on line 1126
```

It's most likely related but I haven't been able to verify that.

In case the steps to replicate are very specific these are the steps I took to get the error:

1. Add 2 subscription products with different billing terms to the cart. 1 gifted and the other not.
2. On checkout enter a recipient email for the non-gifted cart item
3. Attempt to process the checkout. 

https://cloudup.com/ccFBKq7xzHK/f

**Comment by james-allan** [comment]
> I imagine it's related to the fact that adding a recipient on checkout causes the recipient's email to be appended to the recurring cart item key before Subscriptions validates the shipping methods.

I just confirmed this. We update the cart item key on the `woocommerce_checkout_process` action and Subscriptions validates shipping methods on `woocommerce_after_checkout_validation` which is later in the checkout process. This causes the recurring cart key to be different to what Subscriptions expects from the chosen payment methods. 

This potentially raises other issues that I haven't thought about previously. If adding a recipient on check out caused a new recurring cart to be created it will most likely cause issues with shipping method selection for that recurring cart.

**Comment by james-allan** [comment]
fixed with #214
